src/v/storage/offset_translator_state.cc

```diff
@@ -159,30 +194,77 @@ bool offset_translator_state::add_absolute_delta(
           prev, batch_info{.base_offset = base_offset, .next_delta = delta});
         return true;
     } else {
-        int64_t last_delta = _last_offset2batch.rbegin()->second.next_delta;
+        auto [last_offset, last_value] = *_last_offset2batch.rbegin();
+        auto [last_base_offset, last_delta] = last_value;
         int64_t gap_length = delta - last_delta;
+        auto base_offset = offset - model::offset{gap_length};
 
-        if (gap_length != 0) {
-            model::offset last_offset = _last_offset2batch.rbegin()->first;
-            auto base_offset = offset - model::offset{gap_length};
-            vassert(
-              base_offset > last_offset && base_offset < offset,
-              "ntp {}: inconsistent add_absolute_delta (offset {}, delta {}), "
-              "but last_offset: {}, last_delta: {}",
-              _ntp,
-              offset,
-              delta,
+        if (prev > last_offset) {
+            vlog(
+              stlog.debug,
+              "add_absolute_delta: last record: key: {} - {{ bo: {}, delta: {} "
+              "}}",
               last_offset,
+              last_base_offset,
               last_delta);
 
+            if (gap_length < 0) {
+                // This may happen if the delta produced by add_gap calls
+                // is not consistent with add_absolute_delta call.
+                throw std::runtime_error(fmt_with_ctx(
+                  fmt::format,
+                  "ntp {}: inconsistent add_absolute_delta (offset {}, delta "
+                  "{}), "
+                  "but last_offset: {}, last_delta: {}, last_base_offset: {}, "
+                  "gap_length: {}",
+                  _ntp,
+                  offset,
+                  delta,
+                  last_offset,
+                  last_delta,
+                  last_base_offset,
+                  gap_length));
+            }
+
             _last_offset2batch.emplace(
               prev,
               batch_info{.base_offset = base_offset, .next_delta = delta});
             return true;
         } else {
-            return false;
+            // The record batch is already added, the delta computed using
+            // previous calls to 'add_gap' and 'add_absolute_delta' should match
+            // the parameter of this call. There is no need to change the state
+            // of the offset_translator_state. The invariant: any gap above the
+            // one which we're adding should have larger delta.
+            auto it = _last_offset2batch.lower_bound(prev);
+            auto end = _last_offset2batch.end();
+            if (it == end || it->second.next_delta < delta) {
+                vassert(
+                  it != _last_offset2batch.end(),
```

Comment:
Nit: the code is a bit strange in that we check `it == end` in the if condition and then assert that it is not end. Probably should be just assert?